packages/environment/develop.js

```diff
@@ -97,56 +92,54 @@ const Develop = {
       ipc.disconnect(ipcNetwork);
     };
 
-    ipc.connectTo(ipcNetwork, connectPath, function() {
-      ipc.of[ipcNetwork].on("destroy", function() {
-        callback(new Error("IPC connection destroyed"));
-      });
-
-      ipc.of[ipcNetwork].on("truffle.ready", function() {
-        callback(null, disconnect);
-      });
-
-      Object.keys(loggers).forEach(function(key) {
-        var log = loggers[key];
-        if (log) {
-          var message = `truffle.${key}.log`;
-          ipc.of[ipcNetwork].on(message, log);
-        }
+    return new Promise((resolve, reject) => {
+      ipc.connectTo(ipcNetwork, connectPath, function() {
+        ipc.of[ipcNetwork].on("destroy", function() {
+          reject(new Error("IPC connection destroyed"));
+        });
+
+        ipc.of[ipcNetwork].on("truffle.ready", function() {
+          resolve(disconnect);
+        });
+
+        Object.keys(loggers).forEach(function(key) {
+          var log = loggers[key];
+          if (log) {
+            var message = `truffle.${key}.log`;
+            ipc.of[ipcNetwork].on(message, log);
+          }
+        });
       });
     });
   },
 
-  connectOrStart: function(options, ganacheOptions, callback) {
-    const self = this;
-
+  connectOrStart: async function(options, ganacheOptions) {
     options.retry = false;
 
     const ipcNetwork = options.network || "develop";
 
     let connectedAlready = false;
 
-    this.connect(
-      options,
-      async function(error, disconnect) {
-        if (error) {
-          await self.start(ipcNetwork, ganacheOptions);
-
-          options.retry = true;
-          self.connect(
-            options,
-            function(error, disconnect) {
-              if (connectedAlready) return;
-
-              connectedAlready = true;
-              callback(true, disconnect);
-            }
-          );
-        } else {
+    return this.connect(options)
+      .then(disconnect => {
+        connectedAlready = true;
+        return {
+          started: false,
+          disconnect
+        };
+      })
+      .catch(async () => {
+        await this.start(ipcNetwork, ganacheOptions);
+        options.retry = true;
+        return this.connect(options).then(disconnect => {
+          if (connectedAlready) return;
           connectedAlready = true;
-          callback(false, disconnect);
-        }
-      }
-    );
+          return {
+            started: true,
+            disconnect
+          };
+        });
+      });
```

Comment:
(continuing from above because GH suggestions still have strange limitations)

```suggestion
      return {
        started: true,
        disconnect
      };
    }
```